### ISSUE TITLE
Download URLs for HTTP or HTTPS differently structured

### ISSUE THREAD

**Issue by Silvanoc** [eclipse/hawkbit#179]
With the merge of [pull 120](https://github.com/eclipse/hawkbit/pull/120) the structure of the JSON providing the download URLs for HTTP has changed the field "download" to "download-http". But nothing changed for HTTPS, meaning that the download URL remains in the field "download".

Is it a desired behavior? What's the rational behind it?

The undesired consequence is that the existence of the field "download" or "download-http" has to be checked before getting the value.

Have a look at the code after [this line](https://github.com/eclipse/hawkbit/pull/120/files#diff-47bc1d2d2668aeac239a89de04bd9a5cR90) to better understand what I mean.

**Comment by michahirsch** [comment]
Hi @Silvanoc 

the intention to introduce a new `rel` field `download-http` was that changing the existing `rel` `download` cannot be broken as it would be an API break when changing the `rel`. Unfortunate when using external URL-patterns (e.g. an external download-server) , the hawkBit-update-server would currently not know if the external download-server would support `HTTPS`. So the `rel` with `download` is kept for backwards compatibly.

If you check the current master you'll find that in the meantime the `org.eclipse.hawkbit.api.ArtifactUrlHandler` introduced the `protocolSupported` method which allows to only render the URLs in the response which the `ArtifactUrlHandler` supports to render. Nevertheless the `rel` name `download` and `download-http` remains. 

The `download-http` should always contain a non HTTPs-URL e.g. for clients which cannot handle `HTTPs` in case it is supported.

see https://github.com/eclipse/hawkbit/blob/1420b9c2929c3457097dbe4c336bcbb9e73182fb/hawkbit-rest-resource/src/main/java/org/eclipse/hawkbit/controller/DataConversionHelper.java#L92-L105

**Comment by Silvanoc** [comment]
I'm sorry, I'm not sure I got it...

Is this what you mean?
- Download offered via `HTTPS` => HTTPS URL in the field `download`
- Download offered via `HTTP` => HTTP URL in the field `download-http`

If none of them has been activated (`hawkbit.artifact.url.https.enabled=false` and `hawkbit.artifact.url.http.enabled=false`), then the artifacts are listed, but no download URL will be offered. Wouldn't it be more meaningful to return an error? Or is there any use-case that I oversee where such a behavior might be desired?

**Comment by michahirsch** [comment]
Yes that's the current behavior as you explained with the `HTTPS` => `download` and `HTTP` => `download-http`

Not sure if hawkBit should through an error if none is supported. We've also seen clients which are actually not using the URLs rendered from the hawkBit server, instead there are building statically URLs on the client itself. This is definitely not a good approach in my opinion to define statically URLs on the client side but I've seen it and in that case hawkBit don't need to provide any URL at all for downloading. 

So I guess I would avoid throw an exception in case none is supported but I understand that a misconfiguration might happen without noticing.

**Comment by Silvanoc** [comment]
You're right, I never thought on such a use case, but you are right, it's a valid one.

I fully agree with you. I wouldn't throw an exception, but I would at least show a warning when starting Hawkbit or document this behavior as desired in such a use-case.

**Comment by Silvanoc** [comment]
FYI: the API documentation in [this page](http://sp.apps.bosch-iot-cloud.com/documentation/rest-api/softwaremodules-api-guide.html#_get_rest_v1_softwaremodules_softwaremoduleid_artifacts) isn't up-to-date with the behavior you explained in your comments (an is implemented in the code). The field `_links.download-http` is not documented.

**Comment by michahirsch** [comment]
Thanks for the hint, you are right, the documentation is not up-to-date, the field must be described.

Thanks @Silvanoc.

**Comment by Silvanoc** [comment]
I thank you @michahirsch for your prompt reactions! :+1:

I understand that you want the issue to remain open, since you added the label "task". Please feel free to close the issue otherwise.

**Comment by michahirsch** [comment]
@Silvanoc now I recognized that the link to the documentation is directed to the `SoftwareModuleResource` by means this is not the download resource for the `Direct Device Integration API`. I checked the documentation for the `DDI` and there the `download` and `download-http` links are documented as expected :)

The `SoftwareModuleResource` is based on the  `Management API` which I currently does not generate download URLs e.g. for external use to a download-server for devices. 

Here the link to the [Direct Device API Documentation!](http://sp.apps.bosch-iot-cloud.com/documentation/developerguide/apispecifications/directdeviceintegrationapi.html)
